docs/blog/2024-12-04-angular-state-management-2024.md

```diff
@@ -0,0 +1,143 @@
+
+```ts
+@Component({
+  template: `
+    <p>Hello, {{ name() }}</p>
+    <button (click)="updateName()">Update</button>
+  `,
+})
+export class MessageComponent {
+  name = signal('World');
+  constructor() {
+    effect(() => {
+      console.log('Name has changed: ', this.name());
+    });
+  }
+  updateName() {
+    this.name.set('Mike');
+  }
+}
+```
+
+The built-in methods on the `signal` make creating a simple local store very easy to do. If you take Signals and utilize a service, you create your own mini-store.
+
+```ts
+export class AppStore {
+  readonly state = signal([]);
+
+  add(item) {
+    this.state.update((oldState) => [...oldState, item]);
+  }
+  delete(item) {
+    this.state.update((oldState) => oldState.filter((e) => e.id !== item.id));
+  }
+  update(item) {
+    this.state.update((oldState) =>
+      oldState.map((e) => (e.id === item.id ? item : e))
+    );
+  }
+}
+```
+
+Now this is just a very basic approach to managing state, but if you do not want to reach for an additional libray, you could get pretty far with this basic solution. Throw in a few `effect`s in there if need to performs some side effects, and you're golden.
+
+## Signal State
+
+Now if you've been around Angular long enough, you've probably reached for NgRx before, and with good reason. NgRx provides a standard way of managing state in your app that is scalable and testable. In the past, NgRx has provided a store solution based on RxJS, but in more recent release, NgRx has provided a new API based Signals, Signal State and Signal Store.
+
+Signal State is a lightweight API meant to be used in smaller, more isolated scenarios, where a full redux-like API isn't needed. This could be in small to medium sized apps, and in the component itself or extracted to a service.
+
+Reworking our previous example, we can take our signal-based store and update it to use Signal State:
+
+```ts
+import { patchState, signalState } from '@ngrx/signals';
+export class AppStore {
+  readonly state = signalState<Store>({ items: [] });
+
+  addToStore(item: StoreItem) {
+    patchState(this.state, (oldState) => ({
+      ...oldState,
+      items: [...oldState.items, item],
+    }));
+  }
+  removeFromStore(item: StoreItem) {
+    patchState(this.state, (oldState) => ({
+      ...oldState,
+      items: oldState.items.filter((e) => e.id !== item.id),
+    }));
+  }
+  updateStore(item: StoreItem) {
+    patchState(this.state, (oldState) => ({
+      ...oldState,
+      items: oldState.items.map((e) =>
+        e.id === item.id ? { ...item, name: 'bar' } : e
+      ),
+    }));
+  }
+}
+```
+
+Instantly, some things should stand out to you. First, we use the new `signalState` function, instead of the raw `signal` API. Now we can have a more type safe mechanism for interacting with our state. Second, we're no long passing an array. `signalState` only accepts and object/record like value. If you need arrays, you put that on a property of the state.
```

Comment:
```suggestion
Instantly, some things should stand out to you. First, we use the new `signalState` function, instead of the raw `signal` API. Now we can have a more type safe mechanism for interacting with our state. Second, we're no long passing an array. `signalState` only accepts an object/record like value. If you need an array, you put that on a property of the state.
```